### PR TITLE
Clone boards with stacks, labels

### PR DESCRIPTION
**Pull request by jakobroehrl**
* Resolves: #119 
* Target version: vue

* Is there a good Icon I can use?

### CODE REVIEW

File: src/store/main.js [hunk 1]
```diff
@@ -115,6 +115,19 @@ export default new Vuex.Store({
 				state.boards.push(board)
 			}
 		},
+
+		cloneBoard(state, board) {
```
Comment (juliushaertl) [inline src/store/main.js]: I'd say we can call this updateBoard, so we can reue that one later if needed.

**Comment by juliushaertl** [comment]
> Is there a good Icon I can use?

We could use this, but maybe that is a better fit for the move action instead of the clone.
image

But maybe we should just create a simple icon similar to this one:
https://thenounproject.com/search/?q=clone&i=2432924

cc @nextcloud/designers

**Comment by skjnldsv** [comment]
```svg
<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="M4.2 1C3.53 1 3 1.54 3 2.2v1.26h1.2V2.2h9.6v9.6h-1.14V13h1.14c.67 0 1.2-.53 1.2-1.2V2.2c0-.67-.53-1.2-1.2-1.2H4.2z"/></svg>
```

here you go, using the same design as the external icon.
external

File: src/store/main.js (hunk 1 above)
Comment (jakobroehrl) [reply src/store/main.js]: There is already a updateBoard Action Methode. 
Maybe it's confusing if I rename the mutation from clone to update?

**Comment by jakobroehrl** [comment]
@juliushaertl 
Added the Icon, but it is't shown. Can you have a look please?

**Comment by skjnldsv** [comment]
@jakobroehrl please push your work :)

File: css/icons.scss [hunk 2]
```diff
@@ -53,6 +53,10 @@
   background-image: url('../img/toggle-view-collapse.svg');
 }
 
+.icon-clone {
+  background-image: url('../img/clone.svg');
```
Comment (juliushaertl) [inline css/icons.scss]: Please use the icon-black-white mixin similar to line 61:

```
@include icon-black-white('deck', 'clone', 1);
```

**Comment by juliushaertl** [comment]
Looks good in general, do you have debug enabled in your config.php? Otherwise the scss cacher might not update if the underlying scss file has changes.

File: css/icons.scss (hunk 2 above)
Comment (jakobroehrl) [reply css/icons.scss]: thanks, it's working :)

**Comment by jancborchardt** [comment]
> ```
> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="M4.2 1C3.53 1 3 1.54 3 2.2v1.26h1.2V2.2h9.6v9.6h-1.14V13h1.14c.67 0 1.2-.53 1.2-1.2V2.2c0-.67-.53-1.2-1.2-1.2H4.2z"/></svg>
> ```
> 
> here you go, using the same design as the external icon.
> external

Improving the icon a bit, the shapes could be separated to make clear it’s separate, even in small sizes:
```
<svg width="16" height="16" version="1.1" xmlns="http://www.w3.org/2000/svg"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="m4.2 1c-0.67 0-1.2 0.54-1.2 1.2h10.8v10.8c0.67 0 1.2-0.53 1.2-1.2v-9.6c0-0.67-0.53-1.2-1.2-1.2z"/></svg>
```
copy

This is in line with how we do it with other icons like external, icon-menu-sidebar, icon-audio-off and all the -off / -disabled state icons, and similar to how the [Material Design copy icon](https://material.io/resources/icons/?search=copy&icon=file_copy&style=baseline) is made. :)

File: img/clone.svg [hunk 3]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="M4.2 1C3.53 1 3 1.54 3 2.2v1.26h1.2V2.2h9.6v9.6h-1.14V13h1.14c.67 0 1.2-.53 1.2-1.2V2.2c0-.67-.53-1.2-1.2-1.2H4.2z"/></svg>
```
Comment (jancborchardt) [inline img/clone.svg]:
```suggestion
<svg width="16" height="16" version="1.1" xmlns="http://www.w3.org/2000/svg"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="m4.2 1c-0.67 0-1.2 0.54-1.2 1.2h10.8v10.8c0.67 0 1.2-0.53 1.2-1.2v-9.6c0-0.67-0.53-1.2-1.2-1.2z"/></svg>
```

Comment (jakobroehrl) [reply img/clone.svg]: ok, changed

File: css/icons.scss [hunk 4]
```diff
@@ -53,10 +53,15 @@
   background-image: url('../img/toggle-view-collapse.svg');
 }
 
+.icon-clone {
```
Comment (skjnldsv) [inline css/icons.scss]: This is not needed anymore :)

File: img/clone.svg [hunk 5]
```diff
@@ -0,0 +1 @@
+<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.67-.53-1.2-1.2-1.2H2.2C1.53 3 1 3.53 1 4.2v9.6c0 .67.53 1.2 1.2 1.2h9.6c.67 0 1.2-.53 1.2-1.2"/><path d="M4.2 1C3.53 1 3 1.54 3 2.2v1.26h1.2V2.2h9.6v9.6h-1.14V13h1.14c.67 0 1.2-.53 1.2-1.2V2.2c0-.67-.53-1.2-1.2-1.2H4.2z"/></svg>
```
Comment (skjnldsv) [inline img/clone.svg]: Still the wrong icon?
```suggestion
<svg width="16" height="16" xmlns="http://www.w3.org/2000/svg"><path d="M11.8 13.8H2.2V4.2h9.6m1.2 0c0-.7-.5-1.2-1.2-1.2H2.2C1.5 3 1 3.5 1 4.2v9.6c0 .7.5 1.2 1.2 1.2h9.6c.7 0 1.2-.5 1.2-1.2"/><path d="M4.2 1C3.5 1 3 1.5 3 2.2h10.8V13c.7 0 1.2-.5 1.2-1.2V2.2c0-.7-.5-1.2-1.2-1.2z"/></svg>
```

**Comment by jakobroehrl** [comment]
Now it should be fine

**Comment by juliushaertl** [comment]
So some small UX issues:
- [ ] I think we should show a loading indicator until the clone has finished, since this might take some time
- [ ] Ideally the new board starts in editing mode in the sidebar, so you can change the title right away

**Comment by juliushaertl** [comment]
Besides that I'm fine with it :wink: Awesome work so far @jakobroehrl

**Comment by jakobroehrl** [comment]
@juliushaertl 
Loading: done
How can I get the boardId of the just created board? 
newBoard here is empty:
https://github.com/nextcloud/deck/blob/31ae00708d0638daae5ee4296506448f5b502ac6/src/components/navigation/AppNavigationBoard.vue#L137

**Comment by juliushaertl** [comment]
> Loading: done
> How can I get the boardId of the just created board?
> newBoard here is empty:

Does your request return the data?

**Comment by jakobroehrl** [comment]
Yes, it does

File: src/components/navigation/AppNavigationBoard.vue [hunk 6]
```diff
@@ -133,8 +133,14 @@ export default {
 				actions.push({
 					action: () => {
 						this.hideMenu()
-						// this.boardApi.cloneBoard(this.board)
-						this.$store.dispatch('cloneBoard', this.board)
+						this.loading = true
+						this.$store.dispatch('cloneBoard', this.board).then((newBoard) => {
```
Comment (skjnldsv) [inline src/components/navigation/AppNavigationBoard.vue]: please use async await and try, catch :)

**Comment by jakobroehrl** [comment]
> > How can I get the boardId of the just created board?
> > newBoard here is empty:
> 
> Does your request return the data?

@juliushaertl 
Yes, I have the data here: https://github.com/nextcloud/deck/blob/31ae00708d0638daae5ee4296506448f5b502ac6/src/store/main.js#L119

But I need it here:
https://github.com/nextcloud/deck/blob/31ae00708d0638daae5ee4296506448f5b502ac6/src/components/navigation/AppNavigationBoard.vue#L137

File: src/components/navigation/AppNavigationBoard.vue (hunk 6 above)
Comment (jakobroehrl) [reply src/components/navigation/AppNavigationBoard.vue]: Sorry, I don't know what this mean. I'm a Newbie. 
Could you please explain?

File: src/store/main.js (hunk 1 above)
Comment (juliushaertl) [reply src/store/main.js]: Fine then :+1:

File: css/icons.scss [hunk 7]
```diff
@@ -53,10 +53,12 @@
   background-image: url('../img/toggle-view-collapse.svg');
 }
 
+
 @if mixin-exists('icon-black-white') {
 	@include icon-black-white('deck', 'deck', 1);
 	@include icon-black-white('archive', 'deck', 1);
-	@include icon-black-white('circles', 'deck', 1);
+  @include icon-black-white('circles', 'deck', 1);
```
Comment (juliushaertl) [inline css/icons.scss]: Indentation is off

File: lib/Controller/BoardController.php [hunk 8]
```diff
@@ -150,4 +150,13 @@ public function deleteAcl($aclId) {
 		return $this->boardService->deleteAcl($aclId);
 	}
 
+	/**
+	 * @NoAdminRequired
+	 * @param $boardId
+	 * @return \OCP\AppFramework\Db\Entity
```
Comment (juliushaertl) [inline lib/Controller/BoardController.php]: \OCA\Deck\DB\Board is the actual class returned

File: lib/Service/BoardService.php [hunk 9]
```diff
@@ -605,6 +606,49 @@ public function deleteAcl($id) {
...
+
+		if (is_numeric($id) === false) {
+			throw new BadRequestException('board id must be a number');
+		}
+
+		$this->permissionService->checkPermission($this->boardMapper, $id, Acl::PERMISSION_READ);
+		
+		$board = $this->boardMapper->find($id);
+		$newBoard = new Board();
+		$newBoard->setTitle($board->getTitle() . ' (copy)');
```
Comment (juliushaertl) [inline lib/Service/BoardService.php]: `(copy)` should be translatable

Comment (juliushaertl) [inline lib/Service/BoardService.php]:
```suggestion
		$newBoard->setTitle($board->getTitle() . ' (' . $this->l10n->t('copy') . ')');
```

File: src/store/main.js [hunk 10]
```diff
@@ -267,6 +280,15 @@ export default new Vuex.Store({
 					commit('addBoard', board)
 				})
 		},
+		cloneBoard({ commit }, boardData) {
+			return new Promise((resolve, reject) => {
+				apiClient.cloneBoard(boardData)
+					.then((board) => {
+						commit('cloneBoard', board)
+						resolve(board)
+					})
```
Comment (juliushaertl) [inline src/store/main.js]: Please also add a catch block:

```
.catch((err) => {
	return reject(err)
})
```

File: src/services/BoardApi.js [hunk 11]
```diff
@@ -135,6 +135,21 @@ export class BoardApi {
 			})
 	}
 
+	cloneBoard(board) {
+		return axios.post(this.url(`/boards/${board.id}/clone`))
```
Comment (skjnldsv) [inline src/services/BoardApi.js]: please use try catch syntax :)

File: src/components/navigation/AppNavigationBoard.vue [hunk 12]
```diff
@@ -130,6 +130,26 @@ export default {
...
+						this.loading = true
+						try {
+							const newBoard = await this.$store.dispatch('cloneBoard', this.board)
+							
+							this.loading = false
+							const route = this.routeTo
+							route.params.id = newBoard.id
+							this.$router.push(route)
+						}
+						catch {
```
Comment (skjnldsv) [inline src/components/navigation/AppNavigationBoard.vue]:
```suggestion
						} catch(error) {
							console.error(error)
```

File: src/services/BoardApi.js (hunk 11 above)
Comment (jakobroehrl) [reply src/services/BoardApi.js]: Could you give me a hint, how?

Comment (skjnldsv) [reply src/services/BoardApi.js]: Hey! https://javascript.info/async-await :)

Comment (jakobroehrl) [reply src/services/BoardApi.js]: Thanks, now it looks much better! :)

**Pull request merged**